server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt

```diff
@@ -0,0 +1,75 @@
+package org.jetbrains.bsp.bazel.server.sync.languages.python
+
+import ch.epfl.scala.bsp4j.BuildTarget
+import ch.epfl.scala.bsp4j.BuildTargetDataKind
+import ch.epfl.scala.bsp4j.PythonBuildTarget
+import ch.epfl.scala.bsp4j.PythonOptionsItem
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.FileLocation
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.PythonTargetInfo
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo
+import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver
+import org.jetbrains.bsp.bazel.server.sync.BspMappings
+import org.jetbrains.bsp.bazel.server.sync.dependencytree.DependencyTree
+import org.jetbrains.bsp.bazel.server.sync.languages.LanguagePlugin
+import org.jetbrains.bsp.bazel.server.sync.model.Module
+import java.net.URI
+import java.nio.file.Path
+import kotlin.io.path.toPath
+
+class PythonLanguagePlugin(
+    private val bazelPathsResolver: BazelPathsResolver
+) : LanguagePlugin<PythonModule>() {
+
+    override fun resolveModule(targetInfo: TargetInfo): PythonModule? =
+        targetInfo.pythonTargetInfo?.run {
+            PythonModule(
+                calculateInterpreterURI(this),
+                version.takeUnless(String::isNullOrEmpty)
+            )
+        }
+
+    private fun calculateInterpreterURI(pythonTargetInfo: PythonTargetInfo): URI? =
+        pythonTargetInfo.run {
+            interpreter?.let {
+                it.takeUnless { it.relativePath.isNullOrEmpty() }
+                    ?.let { bazelPathsResolver.resolveUri(it) }
+            }
+        }
+
+
+    override fun applyModuleData(moduleData: PythonModule, buildTarget: BuildTarget) {
+        buildTarget.dataKind = BuildTargetDataKind.PYTHON
+        val interpreter = moduleData.interpreter?.toString()
+        buildTarget.data = PythonBuildTarget(moduleData.version, interpreter)
+    }
+
+    fun toPythonOptionsItem(module: Module, pythonModule: PythonModule): PythonOptionsItem =
+        PythonOptionsItem(
+            BspMappings.toBspId(module),
+            emptyList(),
+        )
+
+    override fun dependencySources(targetInfo: TargetInfo, dependencyTree: DependencyTree): Set<URI> =
+        targetInfo.pythonTargetInfo?.run {
+            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.id)
+                .flatMap(::getExternalSources)
+                .map(::calculateExternalSourcePath)
+                .toSet()
+        }.orEmpty()
+
+
+    private fun getExternalSources(targetInfo: TargetInfo): List<FileLocation> =
+        targetInfo.sourcesList.mapNotNull { it.takeIf { it.isExternal } }
+
+    private fun calculateExternalSourcePath(externalSource: FileLocation): URI {
+        val path = bazelPathsResolver.resolveUri(externalSource).toPath()
+        return bazelPathsResolver.resolveUri(findExternalSubdirectory(path) ?: path)
+    }
+
+    private fun findExternalSubdirectory(path: Path?): Path? {
+        return if (path == null || path.parent.endsWith("external"))
+            path
```

Comment:
if it's null u want to return null or it's a bug?

Reply:
we do, cause if the path does not contain site-packages somehow, then the function calling findSitePackagesSubdirectory is going to return the full path

Reply:
okey then let's rewrite it using `when` and make it a bit more explicite + add a tail recursion

Reply:
it is a tail recursion though, it's the equivalent of removing the last directory from the path up to "site-packages"

Reply:
yes yes, u just need to mark it as tailrec to allow compiler do the magic: https://kotlinlang.org/docs/functions.html#tail-recursive-functions